app/src/main/kotlin/compiler/semantic_analysis/VariablePropertiesAnalyzer.kt

```diff
@@ -1,15 +1,105 @@
+import compiler.common.semantic_analysis.ReferenceHashMap
 import compiler.common.semantic_analysis.ReferenceMap
 
 object VariablePropertiesAnalyzer {
-    data class VariableProperties(val owner: Function, val usedInNested: Boolean)
+    data class VariableProperties(val owner: Function?, val usedInNested: Boolean)
 
-    fun calculateVariableProperties(ast: Program, nameResolution: ReferenceMap<Any, NamedNode>, diagnostics: Diagnostics): ReferenceMap<Any, VariableProperties> {
-        return TODO()
+    fun calculateVariableProperties(
+        ast: Program,
+        nameResolution: ReferenceMap<Any, NamedNode>,
+        diagnostics: Diagnostics
+    ): ReferenceMap<Any, VariableProperties> {
+        val variableProperties: MutableReferenceMap<Any, VariableProperties> = ReferenceHashMap()
+        fun resolve(
+            // Any = Statement.Assignment | Expression.Variable
+            node: Any,
+            currentOwner: Function?,
+            isAssignment: Boolean
+        ) {
+            val resolvedVariable: Any = nameResolution.get(node)!!
+            // we assume variable is defined before being referenced
+            // if this fails then something is wrong with name resolution
+            val oldOwner = variableProperties.get(resolvedVariable)!!.owner
+            if (resolvedVariable is Function.Parameter && isAssignment) {
+                // function parameter must have an owner
+                diagnostics.report(AssignmentToFunctionParameter(resolvedVariable, oldOwner!!, currentOwner!!))
+            }
+            if (oldOwner != currentOwner) {
+                // if this was a function that is not nested in the original owner
+                // then name resolution would not be able to succeed
+                // current owner cannot be null as it is always a function once we start descending
+                if (isAssignment)
+                    diagnostics.report(AssignmentToOuterVariable(resolvedVariable, oldOwner, currentOwner!!))
+                variableProperties.put(resolvedVariable, VariableProperties(oldOwner, true))
+            }
+        }
+
+        // Any = Expression | Statement
+        fun analyzeVariables(node: Any, currentOwner: Function?) {
+            when (node) {
+                is Statement.Evaluation -> analyzeVariables(node.expression, currentOwner)
+                is Statement.VariableDefinition -> {
+                    variableProperties.put(
+                        node.variable,
+                        VariableProperties(currentOwner, false)
+                    )
+                    node.variable.value?.let { analyzeVariables(it, currentOwner) }
+                }
+                is Global.VariableDefinition -> {
+                    variableProperties.put(
+                        node.variable,
+                        VariableProperties(currentOwner, false)
+                    )
+                    node.variable.value?.let { analyzeVariables(it, currentOwner) }
```

Comment:
To avoid code repetition in `Statement.VariableDefinition` and `Global.VariableDefinition`, maybe we could handle this the same way as functions?
```
is Statement.VariableDefinition -> analyzeVariables(node.variable, currentOwner)
is Global.VariableDefinition -> analyzeVariables(node.variable, currentOwner)
...
is Variable-> {
   //common code
}
```